### PR TITLE
Ridge Chapter

### PR DESCRIPTION
**Pull request by hschafer**
I did a horrible thing and this branch has a ton of work on it. I kept going back and forth on content work and web-dev work (since I was updating the things that I needed as I wrote the Ridge chapter). I went too far into development and then had difficulty separating the commits. This PR is HUGE as a result 😞. 

I wouldn't spend too much time looking at the code itself and would probably spend more of your efforts on the actual content in the ridge chapter and maybe exploring the new features outlined below to make sure they look right in the browser to you.

At a high level, the changes are:
* The draft of the Ridge reading 😊 
* Update styling and some behavior of the test your understanding cards, including the ability to expand/close all of them.
   * Also includes CSS to allow margin notes inside TYU cards.
* To support the last point, added some new code to handle React contexts for that. While I was at it, I cleaned up some of that MarginNote code too.
   * To this point, this involves changing pretty much every reading file to no longer pass those counters around.
* Some styling changes for Latex to make it look better. Made display math text larger and inline math line-wrap on small screens.
* Cleaned up some of the case study logic for the nav bar. Included some styling changes to fix a bug where clicking on the edges of the dropdown just closed it without doing anything. 
* Some minor global CSS changes to fix up some madness about width setting that we inherited from Tufte. The bug that made me spot this was the inline math `\\hat{y}` was misplaced. I think I fixed it so it's a bit more reasonable default and should cause fewer problems for new additions.

I'M SO SORRY 😭

Also happy to walk you through any of these changes if you do end up looking at code.

### CODE REVIEW

**Comment by hschafer** [comment]
I'm so sorry. I did a horrible thing and this branch has a ton of work on it. I went too far into development and then had difficulty separating the commits. This PR is HUGE so I wouldn't spend too much time looking at the code and focus more checking out the behavior and the text for the ridge chapter. 

At a high level, the changes are:
* The draft of the Ridge reading 😊 
* Update styling and some behavior of the test your understanding cards, including the ability to expand/close all of them.
   * Also includes CSS to allow margin notes inside TYU cards.
* To support the last point, added some new code to handle React contexts for that. While I was at it, I cleaned up some of that MarginNote code too.
   * To this point, this involves changing pretty much every reading file to no longer pass those counters around.
* Some styling changes for Latex to make it look better. Made display math text larger and inline math line-wrap on small screens.
* Cleaned up some of the case study logic for the nav bar. Included some styling changes to fix a bug where clicking on the edges of the dropdown just closed it without doing anything. 
* Some minor global CSS changes to fix up some madness about width setting that we inherited from Tufte. The bug that made me spot this was the inline math `\\hat{y}` was misplaced. I think I fixed it so it's a bit more reasonable default and should cause fewer problems for new additions.

I'M SO SORRY.

File: website/pages/chapters/ridge.js [hunk 1]
```diff
@@ -0,0 +1,845 @@
...
+          can convince yourself that this will also solve the cancelling-out
+          problem like the L1 norm. In the next chapter, we will see how the
+          differing effects of the L1 and L2 norms, but for the rest of this
+          chapter, we will focus on this L2 norm.
+        </p>
+
+        <h3>Ridge Regression</h3>
+        <p>
+          When we use regularization in our regression context with the L2 norm,
+          our quality metric then becomes the <b>ridge regression</b> quality
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:518]: do we have any intuition for why it's called ridge? might be helpful to associate this with the l2 norm. I don't remember any, it only took me 3 ml classes to commit ridge and lasso to memory lol.

File: website/pages/chapters/ridge.js [hunk 2]
```diff
@@ -0,0 +1,845 @@
...
+              <IM math={`\\lambda||x||_2^2 \\rightarrow \\infty`} /> since{" "}
+              <IM math={`||x||_2^2 > 0`} />. Even if that magnitude is small,
+              you can blow up that penalty by choosing a super large{" "}
+              <IM math={`\\lambda`} />.
+            </p>
+
+            <p>
+              Consider a different case where every coefficient{" "}
+              <IM math={`j`} />, <IM math={`w_j = 0`} />. We would say then that
+              the whole vector <IM math={`w = 0`} /> as a notational
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:616]: consider boldfacing the 0 or adding a vector arrow over the 0?

File: website/pages/chapters/ridge.js [hunk 3]
```diff
@@ -0,0 +1,845 @@
...
+          <IM math={`\\lambda`} />. Like in our discussion before, you can
+          imagine that if you let <IM math={`\\lambda \\rightarrow \\infty`} />,
+          all of the coefficients will be 0. Each vertical-line of circles
+          represents a single ridge regression predictor (trained with that{" "}
+          <IM math={`\\lambda`} />
+          ). Each feature's coefficient is associated a color, and you can see
+          how that coefficient changes as you increase <IM math={`\\lambda`} />.
+        </p>
+
+        <figure className="fullwidth">
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:655]: seems weird that the image bleeds into the footnotes column, but this is just a personal preference.

File: website/pages/chapters/ridge.js [hunk 4]
```diff
@@ -0,0 +1,845 @@
...
+          sure it works well.
+        </p>
+
+        <p>
+          The first is a rather subtle point about which coefficients you should
+          regularize. In our explanation above, we showed that you should
+          regularize the whole coefficient vector{" "}
+          <IM math={`w = [w_0, w_1, ..., w_D]`} />. That might have been a bit
+          overkill and will likely cause problems in practice{" "}
+          <MarginNote>
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:694]: you need an edgier meme :p

File: website/pages/chapters/ridge.js [hunk 5]
```diff
@@ -0,0 +1,845 @@
...
+            />
+          </li>
+          <li>
+            Center the data so the <IM math={`y`} /> values are have mean 0. If
+            this is the case, then favoring a <IM math={`w_0`} /> is not a
+            problem since the data is centered at 0.
+          </li>
+        </ul>
+
+        <p>
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:741]: I'd add an `<h3>` tag called "Normalization" as that's a pretty important practical detail.

Comment (Rockett8855) [reply website/pages/chapters/ridge.js:741]: And in the previous bullet id introduce the term normalization, so it's crystal clear you're talking about normalization in the previous section.

File: website/pages/chapters/ridge.js [hunk 6]
```diff
@@ -0,0 +1,845 @@
...
+        </p>
+        <BM
+          math={`\\sigma_j(Train) = \\sqrt{\\frac{1}{|Train|} \\sum_{x\\in Train}\\left( h_j(x) - \\mu_j(Train) \\right)^2}`}
+        />
+
+        <p>
+          So by pre-processing our data to be normalized, we don't have to worry
+          about features with wildly different scales being falsely penalized by
+          the regularization term.
+        </p>
```
Comment (Rockett8855) [inline website/pages/chapters/ridge.js:787]: Maybe add a practice problem of doing normalization. i know when I was learning this stuff, I felt like regularization was a fuzzy concept. IDK the best way to do it, but maybe make them calculated the normalization by hand, and give them a learned model, so they can see the cofficient values? ALSO while you don't normalize y (it looks like) I feel like people not only normalize x, but y as well?

**Comment by hschafer** [comment]
> I pretty much only read the ridge chapter, the css and js look pretty trivial, so i trust ya.
> 
> The only question I have overall is why would you want a feature to "open all" test your understanding cards, I feel like you're just tempting them to do it instead of going through the exercise of trying to answer the question and clicking the answers.

That's a good point. I think the risk of that is outweighed by the benefit this allows of making **review** easier. For example, I put a margin note and a bold term in one of those boxes. If I didn't have that button, it would be harder for people to review if they miss one of those terms or margin notes. I think regardless, it's up to the student to self-regulate to think of the answer before clicking and adding the expand all won't change their behavior (I think).

File: website/pages/chapters/ridge.js [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+          </TYU.Explanation>
+        </TYU>
+
+        <p>
+          When using multiple features (commonly called Multiple Linear
+          Regression), you can also usually interpret the coefficients in a
+          similar manner. Suppose for our housing example, we want to learn a
+          predictor of the form{" "}
+          <IM
+            math={`\\hat{y} = \\hat{w}_0 + \\hat{w}_1x[1] + \\hat{w}_2x[2]`}
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:144]: Ah, in our notation earlier we differentiate that `x_i` means the ith example from your dataset while `x[j]` is the jth data input for an example. I try as much as possible to avoid `x_{i,j}` since it's awfully confusing for newbies.

In fact, it's actually kind of wack that we use `w_0` instead of `w[0]`. Maybe something to fix in the future @_@

File: website/pages/chapters/ridge.js [hunk 8]
```diff
@@ -0,0 +1,845 @@
...
+          <em>
+            it is under the condition that all other features stay fixed in
+            value
+          </em>
+          . So in our housing example, <IM math={`\\hat{w}_1`} /> is the
+          increase in price per increase in square footage,{" "}
+          <em>holding the number of bathrooms constant</em>. It is a bit harder
+          to try to interpret the change in price if you vary both the size and
+          number of bathrooms. This interpretation is only valid if you hold all
+          the other features constant.
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:167]: Do you mean something like 

----

Suppose your ML algorithm learned the predictor `\hat{y} = 100 + 200x[1] + 300x[2]`. Using this discussion, you would interpret the learned coefficients as follows:
* `\hat{w}_0 = 100` means the price of a 0 square foot house with 0 bathrooms
* `\hat{w}_1 = 200` means if you add one square foot to the house, the price should increase by 200, *assuming the number of bathrooms do not change*.
* `\hat{w}_2 = 300` means if you add one more bathroom to the house, the price should increase by 300, *assuming the square footage does not change*.

File: website/pages/chapters/ridge.js [hunk 9]
```diff
@@ -0,0 +1,845 @@
...
+        <p>
+          A consequence of this need to fix all other features, makes something
+          like the coefficients in polynomial regression difficult to interpret.
+          If you learn a predictor <IM math={`\\hat{y} = 2 + 3x + 4x^2`} />,
+          it's a little more complicated to say how the output changes per
+          unit-change in <IM math="x" /> since you can't really fix the other
+          terms.
+        </p>
+
+        <h3>A note on interpretability</h3>
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:187]: Yay! 😄

File: website/pages/chapters/ridge.js [hunk 10]
```diff
@@ -0,0 +1,845 @@
...
+          tend to have really large coefficients as they become more and more
+          overfit. In other words, in order to get that really wiggly behavior
+          in the polynomial, the coefficients must be large in magnitude (
+          <IM math={`|\\hat{w}_j| >> 0`} />
+          ). The following animation shows how increasing the degree{" "}
+          <IM math="p" /> of a polynomial regression model tends to increase the
+          magnitudes of the coefficients.
+        </p>
+
+        <p>TODO(manim) coefficient growth?</p>
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:319]: Sorry, I don't think I quite follow what you mean here. I think we established why these wiggly polynomials are bad in the Assessing Performance chapter so I'm not sure if we need another animation to show that it's bad again. 

I should have maybe written a clearer note for what I meant here since this is extremely vague. For this animation, I was intending to show that as you increase the degree `p`, the magnitudes of the coefficients increase. There is a plot of this in that Jupyter notebook I made where you can see that the coefficients get HUGE even for moderate `p`.

File: website/pages/chapters/ridge.js [hunk 11]
```diff
@@ -0,0 +1,845 @@
...
+
+        <p>
+          In the last chapter, we saw that we could use cross validation or a
+          validation set to pick which model complexity to use. For the case of
+          polynomial regression and choosing the degree <IM math={`p`} />, this
+          is relatively simple since we just need to try each degree once. If we
+          are considering multiple linear regression that needs to choose from{" "}
+          <IM math={`d`} /> features, there are many models we will need to
+          evaluate. In fact, with <IM math={`d`} /> possible features to select
+          from, there are <IM math={`2^d`} /> models that use every possible
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:359]: Oh, I didn't mean to bring up interaction features at all here. Is there something I could reword to not evoke that idea? I was just saying if you were trying to try a model using every subset of the available data inputs, there will be $2^d$ of them.

In an earlier chapter, we showed an example of a feature transformation that made an interaction term but did not dwell on it.

File: website/pages/chapters/ridge.js [hunk 12]
```diff
@@ -0,0 +1,845 @@
...
+
+        <p>
+          We combine both of these values into a new quality metric that our ML
+          algorithm will minimize.
+        </p>
+
+        <BM math={`\\hat{w} = \\min_w L(w) + \\lambda R(w)`} />
+
+        <p>
+          <IM math={`\\lambda`} /> will be some tuning parameter that lets us
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:426]: I am torn on this one. The $\lambda$ is kind of a weird idea we are introducing already and I'm worried about overloading that concept with a new term. I added it in the recap section originally lol. I could be convinced to just throw the discussion of the term here (or maybe put it in a margin note?).

File: website/pages/chapters/ridge.js [hunk 13]
```diff
@@ -0,0 +1,845 @@
...
+        <p>
+          When we add regularization, we change our quality metric to include
+          both:
+        </p>
+        <ul>
+          <li>
+            <IM math={`L(w)`} /> The original loss function of interest.
+          </li>
+          <li>
+            <IM math={`R(w)`} /> A new measurement of the magnitude of the
```
[the remark this answers is not in the document]
Comment (hschafer) [reply website/pages/chapters/ridge.js:413]: Let me know what you think now.

File: website/pages/chapters/ridge.js (hunk 1 above)
Comment (hschafer) [reply website/pages/chapters/ridge.js:518]: From Wikipedia: 

>  In statistics, the method is known as ridge regression, in machine learning it is known as weight decay, and with multiple independent discoveries, it is also variously known as the Tikhonov–Miller method, the Phillips–Twomey method, the constrained linear inversion method, and the method of linear regularization. It is related to the Levenberg–Marquardt algorithm for non-linear least-squares problems.

Comment (hschafer) [reply website/pages/chapters/ridge.js:518]: Ah it looks like it is trying to fix the "ridge" problem where you have a co-linearity in your data and there is no unique global minimum to the RSS. 

image

File: website/pages/chapters/ridge.js (hunk 3 above)
Comment (hschafer) [reply website/pages/chapters/ridge.js:655]: I could go back to it being in the main column. The Tufte style allows these fullwidth images to provide more detail. I kind of like how it looks, but if it bothers you I can change it.

File: website/pages/chapters/ridge.js (hunk 5 above)
Comment (hschafer) [reply website/pages/chapters/ridge.js:741]: Hmmm, technically you don't need to normalize the data for the intercept problem. I've always thought of normalization as mean centering and scaling. You don't need to change the scale of the `ys` for the intercept problem, just make sure they are mean 0.

File: website/pages/chapters/ridge.js (hunk 6 above)
Comment (hschafer) [reply website/pages/chapters/ridge.js:787]: I think a practice problem for regularization is a great idea that I'll probably save for class. I have left out most of the poll everywhere questions from the slides because I want to use those for in-person activities. I have to have SOMETHING to do in class 😛 

I added a note about normalizing y's. It is common in practice, but not strictly necessary. The big problem with the inputs is having them on _different_ scales. If you made them all mean 50 std deviation 47, it would be fine. The problem comes from having one feature that requires large coefficients and one that requires small since then you don't penalize them the same.

File: website/pages/chapters/ridge.js (hunk 2 above)
Comment (hschafer) [reply website/pages/chapters/ridge.js:616]: I'll boldface to avoid the arrow notation. Not that it's bad, but it's just another notational thing I want to avoid if strictly necessary (and I mention the `w=0` is a notational convenience).

**Comment by hschafer** [comment]
I think I addressed all your comments and I left a few unresolved if I didn't think I had a good answer yet.

File: website/pages/chapters/ridge.js (hunk 3 above)
Comment (Rockett8855) [reply website/pages/chapters/ridge.js:664]: I think for personal preference comments, i'll just leave them when i see something i think should change, and you should just be the final decision :P. If i feel very strongly about something I won't tag is as a personal preference.

File: website/pages/chapters/ridge.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Rockett8855) [reply website/pages/chapters/ridge.js:320]: Coming back I don't really know what i was trying to get at either.

**Comment by Rockett8855** [comment]
Just noticed that you added the term hyperparameter in the recap. I think it's kinda weird to introduce a concept/term in the recap. maybe not?

**Comment by hschafer** [comment]
> Just noticed that you added the term hyperparameter in the recap. I think it's kinda weird to introduce a concept/term in the recap. maybe not?

Yep, that is a bit weird. I moved it to the margin note since we will start Chapter 4 explaining that terminology right away.
